source/h2_connection.c

```diff
@@ -715,6 +919,44 @@ static void s_stream_complete(struct aws_h2_connection *connection, struct aws_h
     aws_http_stream_release(&stream->base);
 }
 
+int aws_h2_connection_on_stream_closed(
+    struct aws_h2_connection *connection,
+    struct aws_h2_stream *stream,
+    enum aws_h2_stream_closed_when closed_when,
+    int aws_error_code) {
+
+    AWS_PRECONDITION(aws_channel_thread_is_callers_thread(connection->base.channel_slot->channel));
+    AWS_PRECONDITION(stream->thread_data.state == AWS_H2_STREAM_STATE_CLOSED);
+    AWS_PRECONDITION(stream->base.id != 0);
+
+    uint32_t stream_id = stream->base.id;
+
+    /* Mark stream complete. This removes the stream from any "active" datastructures,
+     * invokes its completion callback, and releases its refcount. */
+    s_stream_complete(connection, stream, aws_error_code);
+    stream = NULL; /* Reference released, do not touch again */
+
+    /* Peer might have already sent/queued frames for this stream before learning that we closed it.
+     * But if peer was the one to close the stream via RST_STREAM, they know better than to send more frames. */
+    bool frames_might_trickle_in = closed_when != AWS_H2_STREAM_CLOSED_WHEN_RST_STREAM_RECEIVED;
+    if (frames_might_trickle_in) {
+        if (aws_hash_table_put(
+                &connection->thread_data.closed_streams_where_frames_might_trickle_in,
+                (void *)(size_t)stream_id,
+                (void *)(size_t)closed_when,
+                NULL)) {
+
+            AWS_H2_STREAM_LOG(ERROR, stream, "Failed inserting ID into map of recently closed streams");
+            return AWS_OP_ERR;
+        }
+
+        /* #TODO remove entries from closed_streams_where_frames_might_trickle_in after some period of time */
```

Comment:
Wait so are they in there indefinitely right now?

Reply:
yes
I was trying to keep down the PR size so didn't tackle it yet